### PR TITLE
[CO-342] Backward-compatiblity parser for V0 CBackupPhrase

### PR DESCRIPTION
**Pull request by Anviking**
## Linked issue

[CO-342](https://iohk.myjetbrains.com/youtrack/issue/CO-342#tab=Comments)

## Type of change

- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [~] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [~] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: wallet/src/Pos/Wallet/Aeson/ClientTypes.hs [hunk 1]
```diff
@@ -53,15 +54,26 @@ instance ToJSON (CBackupPhrase mw) where
     toJSON =
         genericToJSON defaultOptions
 
+backwardsCompitability :: [Text] -> [Text]
+backwardsCompitability mnemonicWords = mnemonicWords >>= splitWords
+    where
+        splitWords :: Text -> [Text]
+        splitWords = split (== ' ')
```
Comment (akegalj) [inline wallet/src/Pos/Wallet/Aeson/ClientTypes.hs]: why not https://hackage.haskell.org/package/text-1.2.3.0/docs/Data-Text.html#v:words ?

Comment (Anviking) [reply wallet/src/Pos/Wallet/Aeson/ClientTypes.hs]: 👍 thanks!

File: wallet/src/Pos/Wallet/Aeson/ClientTypes.hs [hunk 2]
```diff
@@ -53,15 +53,36 @@ instance ToJSON (CBackupPhrase mw) where
...
-    parseJSON =
-        genericParseJSON defaultOptions
+    parseJSON (Object v) =
+        (v .: "bpToList")
+        <&> backwardCompitability
+        >>= (eitherToParser . mkMnemonic)
+        <&> CBackupPhrase
+      where
+        backwardCompitability :: [Text] -> [Text]
+        backwardCompitability = (>>= words)
```
Comment (KtorZ) [inline wallet/src/Pos/Wallet/Aeson/ClientTypes.hs]: ¯\\\_(ツ)_/¯ Com**pita**bility ~> _Compatibility_

File: wallet/test/Test/Pos/Util/MnemonicSpec.hs [hunk 3]
```diff
@@ -77,6 +77,11 @@ spec = do
         it "CBackupPhrase from JSON" $ forM_ testVectors $ \(bytes, _, mnemonic, _, _) ->
             Aeson.decode (jsonV0Compat bytes) `shouldBe` pure (CBackupPhrase mnemonic)
 
+        it "CBackupPhrase from legacy JSON" $
+            (shouldBe @(Maybe (CBackupPhrase 12)))
+            (Aeson.decode legacy)
+            (Aeson.decode modern)
+
```
[the remark this answers is not in the document]
Comment (KtorZ) [reply wallet/test/Test/Pos/Util/MnemonicSpec.hs]: Indeed. Since we already have the `testVectors`, I'd suggest to add an element to the tuple with the `legacy` version of it, and perform a test similar to the above:

```hs
it "CBackupPhrase from legacy JSON" $ forM_ testVectors $ \(_, legacyBytes_, mnemonic, _, _) ->
    Aeson.decode (jsonV0Compat legacyBytes) `shouldBe` pure (CBackupPhrase mnemonic)

it "Mnemonic from legacy JSON" $ forM_ testVectors $ \(_, legacyBytes_, mnemonic, _, _) ->
    Aeson.decode legacyBytes `shouldSatisfy` isLeft  
```

[the remark this answers is not in the document]
Comment (KtorZ) [reply wallet/test/Test/Pos/Util/MnemonicSpec.hs]: Also, think that the parser use to accept any combination of space-separated words. So:

```
["abandon", "abandon abandon", "abandon abandon abandon abandon abandon", "abandon abandon abandon abandon"]

["abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon "]
```

were both valid inputs. Try to reflect that aspect in the test vectors you add :+1:

File: wallet/src/Pos/Util/Mnemonic.hs [hunk 4]
```diff
@@ -348,10 +349,11 @@ instance
     ) => FromJSON (Mnemonic mw) where
     parseJSON =
         parseJSON >=> (eitherToParser . mkMnemonic)
-      where
-        eitherToParser :: Either MnemonicErr (Mnemonic mw) -> Parser (Mnemonic mw)
-        eitherToParser =
-            either (fail . formatToString build) pure
+
+-- TODO: Mnemonic can be made generic. Can MnemonicErr?
```
Comment (KtorZ) [inline wallet/src/Pos/Util/Mnemonic.hs]: I don't see any obvious reasons why it couldn't.

Comment (KtorZ) [reply wallet/src/Pos/Util/Mnemonic.hs]: And we might leverage that in the parser.

File: wallet/src/Pos/Wallet/Aeson/ClientTypes.hs [hunk 5]
```diff
@@ -53,15 +53,36 @@ instance ToJSON (CBackupPhrase mw) where
...
 
+
+-- We use a custom parser to support
+-- {
+--   "bpToList": ["derive pulse priority impose execute balcony guard hidden
+--    snap unhappy alcohol next"]
+-- }
+-- as in addition to
+-- {
+--    "bpToList": ["derive", "pulse", "priority", "impose", "execute",
```
Comment (adinapoli-iohk) [inline wallet/src/Pos/Wallet/Aeson/ClientTypes.hs]: @KtorZ @Anviking lately I have become paranoid in mentioning _any_ valid 12 word mnemonic in our source code.

If possible, I would be inclined to either not have them here or to have dummy ones (i.e. a selection of 12 random words not suitable for BIP39) 😉

File: wallet/test/Test/Pos/Util/MnemonicSpec.hs [hunk 6]
```diff
@@ -130,3 +135,11 @@ spec = do
         $ Aeson.eitherDecode
         $ Aeson.encode
         $ def @(Mnemonic 12)
+
+    -- | Sample json for testing backward compatibility edge case
+    legacy :: BL.ByteString
+    legacy = "{\"bpToList\":[\"derive pulse priority impose execute balcony guard hidden snap unhappy alcohol next\"]}"
```
Comment (adinapoli-iohk) [inline wallet/test/Test/Pos/Util/MnemonicSpec.hs]: Ditto the above ^

I am a bit wary about having hardcoded mnemonics, I would like to avoid some curious user using these for his wallet.

Perhaps you could use the generator to generate a valid 12 word mnemonic and then use those in this test?

It's pretty much the same thing I did in my [specs](https://github.com/input-output-hk/cardano-sl/blob/develop/wallet-new/test/unit/Test/Spec/CreateWallet.hs#L57) as well.

Comment (KtorZ) [reply wallet/test/Test/Pos/Util/MnemonicSpec.hs]: Agreed, even though, for the test vectors, I'd rather keep them as it is. I hardly think a lot of users / exchanges read the source code and it's even less likely they read tests. 

Another option is simply to extend the list of "forbidden" mnemonics with all the one present in the test vectors.

Comment (KtorZ) [reply wallet/test/Test/Pos/Util/MnemonicSpec.hs]: https://github.com/input-output-hk/cardano-sl/blob/develop/wallet/src/Pos/Util/Mnemonic.hs#L213-L224

File: wallet/src/Pos/Wallet/Aeson/ClientTypes.hs (hunk 5 above)
Comment (KtorZ) [reply wallet/src/Pos/Wallet/Aeson/ClientTypes.hs]: Yup :+1: Good one @adinapoli-iohk, let's stick to:

https://github.com/input-output-hk/cardano-sl/blob/develop/wallet/src/Pos/Util/Mnemonic.hs#L313-L324

When showing any example.

File: wallet/test/Test/Pos/Util/MnemonicSpec.hs [hunk 7]
```diff
@@ -65,31 +66,38 @@ spec = do
         it "Can generate 128 bits entropy" $
             (length . entropyToByteString <$> genEntropy @128) `shouldReturn` 16
 
-        it "Mnemonic to JSON" $ forM_ testVectors $ \(bytes, _, mnemonic, _, _) ->
+        it "Mnemonic to JSON" $ forM_ testVectors $ \(bytes, _, _, mnemonic, _, _) ->
```
Comment (Anviking) [inline wallet/test/Test/Pos/Util/MnemonicSpec.hs]: @KtorZ what about record wildcards instead of the `TestVector` tuple?

File: wallet/test/Test/Pos/Util/MnemonicSpec.hs [hunk 8]
```diff
@@ -28,6 +28,7 @@ instance Eq CC.XPrv where
 
 type TestVector =
     ( BL.ByteString             -- ^ Raw JSON encoding (V1)
+    , BL.ByteString             -- ^ Precursor to legacy (V0) JSON encoding
```
Comment (Anviking) [inline wallet/test/Test/Pos/Util/MnemonicSpec.hs]: This is slightly awkward to describe. Could be more verbose if changing the comment format.

Comment (Anviking) [reply wallet/test/Test/Pos/Util/MnemonicSpec.hs]: As there is now no other comment in this file explaining the legacy format, this should probably be more detailed. Think I'll do it.

File: wallet/test/Test/Pos/Util/MnemonicSpec.hs [hunk 9]
```diff
@@ -27,11 +27,25 @@ instance Eq CC.XPrv where
...
+    , Entropy (EntropySize 12)
+
+      -- | Corresponding Mnemonic
+    , Mnemonic 12
+
+      -- | Corresponding Seed
+    , ByteString
+
+      -- | Corresponding AESKey
+    , AesKey
```
Comment (Anviking) [inline wallet/test/Test/Pos/Util/MnemonicSpec.hs]: Can revert this easily (4dd28775d1ac128f697a13c4d57e8ca0cc453c23)

File: wallet/test/Test/Pos/Util/MnemonicSpec.hs (hunk 7 above)
Comment (KtorZ) [reply wallet/test/Test/Pos/Util/MnemonicSpec.hs]: I'd prone lenses over wildcards if anything, but yeah. As always, it started as a tuple with 2 elements aaaaand... now we have this little monster.
Feel free to refactor that part; this will also re-trigger the failing AppVeyor CI :/
